File: 0055-jump-game/0055-jump-game.cpp
class Solution {
public:
    
    bool isreach(vector<int> &arr, int curr, int n, vector<int> &dp){
        if(curr >= n-1) return true;
        if(dp[curr] != -1) return dp[curr] == 0 ? false : true;
        if(arr[curr] == 0){
            dp[curr] = 0;
            return false;
        }
        
        bool check = false;
        for(int i = 1;i<=arr[curr];i++){
            if(isreach(arr, curr+i, n, dp) == true){
                check = true;
                dp[curr] = 1;
                break;
            }
            else dp[curr] = 0;
        }
        dp[curr] = check == false ? 0 : 1;
        return check;
    }
    
    bool canJump(vector<int>& arr) {
        int n = arr.size();
        if(n == 1) return true;
        vector<int> dp(n, -1);
        return isreach(arr, 0, n, dp);
    }
};

File: 0055-jump-game/0055-jump-game.cs
public class Solution {
    public bool CanJump(int[] arr) {
        if(arr.Count() > 1 && arr[0] == 0) return false;

        int maxIdx = 0;

        for(int i = 0; i < arr.Length; i++){
            if(maxIdx >= arr.Count() - 1) return true;
            if(i + arr[i] < maxIdx) continue;
            else{
                if(i > maxIdx) return false;
                maxIdx = i + arr[i];
            }
        }
        return false;
    }
}